codepropertygraph/src/main/scala/io/shiftleft/utils/TimeMetric.scala

```diff
@@ -0,0 +1,349 @@
+package io.shiftleft.utils
+
+import com.sun.management.OperatingSystemMXBean
+import org.slf4j.{Logger, LoggerFactory}
+
+import java.io.PrintWriter
+import java.lang.management.ManagementFactory
+import java.util
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.{Calendar, Date}
+import scala.collection.mutable
+case class TimeMetricRecordConfig(resultFile: String = "cpuandmemoryusage.csv", recordFreq: Int = 60000)
+case class MaxPrintBuffers(MAX_FULLNAME_BUFFER: Int = 90, MAX_NAME_BUFFER: Int = 55, MAX_DONE_TIME_BUFFER: Int = 40)
+object TimeMetric {
```

Comment:
Could this be a `class` instead? It would help to containerize all these variables into a single object instance which can be garbage collected much more easily.

Reply:
Is it the case that multiple threads/passes are writing to this at the same time?

Reply:
+1 for what David said - currently this is global mutable state, which may be accessed and modified by concurrently running CpgPasses.
(yes, currently they run in sequence, but the idea is that we can parallelise independent ones)

Reply:
It is not expected to call this TimeMetric from multiple places. It will be called from two places.
1. Parent stages like AST Creation, Overlay or OSS Dataflows. Which I assume will be mostly called from the main thread. These parent stages are like groups of passes. 
2. Second place it will be called from the passes, before all the tasks are spawned. In principle from their base class either from `ForkJoinParallelCpgPass` or `ConcurrentWriterCpgPass`. It gets before all the tasks are spawned so it will get called again from the main thread.

I will add few more details in the description which will give more clarity on it.